pimd: Extend multicast boundary/ACL functionality

csiltala opened the pull request:
- Add new interface command `ip multicast boundary ACCESSLIST4_NAME`. This allows filtering on both source and group vs. group-only as with the existing `ip multicast boundary oil WORD` command, which uses prefix-lists. If both are configured, the prefix-list is evaluated first.
- Add basic test and documentation
- Add documentation for existing access-list functionality

csiltala committed:
tests: Add basic multicast boundary test

Add simple test to show filtering of IGMP joins using new "ip multicast
boundary" filtering with access-lists, include test of existing prefix-
list based "ip multicast boundary oil" command.

Signed-off-by: Corey Siltala <csiltala@atcorp.com>

csiltala committed:
doc: Expand ACL and multicast boundary documentation

Add documentation for existing extended access-list functionality and
the new "ip multicast boundary" command leveraging that functionality.

Signed-off-by: Corey Siltala <csiltala@atcorp.com>

Jafaral commented on pimd/pim_iface.c:
```diff
@@ -1258,6 +1259,12 @@ static int gm_join_sock(const char *ifname, ifindex_t ifindex,
 {
 	int join_fd;
 
+	if (pim_is_group_filtered(pim_ifp, &group_addr, &source_addr)) {
+		zlog_warn("%s: join failed for (S,G)=(%pPAs,%pPAs) due to multicast boundary filtering",
+			  __func__, &source_addr, &group_addr);
+		return -1;
+	}
```
This should be a debug rather than a warning. It is normal operation.

Jafaral commented on pimd/pim_join.c:
```diff
@@ -312,10 +312,6 @@ int pim_joinprune_recv(struct interface *ifp, struct pim_neighbor *neigh,
 
 			buf += addr_offset;
 
-			/* if we are filtering this group, skip the prune */
-			if (filtered)
-				continue;
-
```
Is the reasoning here is that we should always scan the prune sources regardless of filtering rules?

csilt replied:
Prunes for flows for which we have no state have no effect, but if we begin filtering a (S,G) that is already being forwarded, then we'll want to continue processing prunes for that flow if it tries to prune before the flow state times out.

donaldsharp commented:
Is there any particular reason you don't actually have any commit message?  Make it easier for the person coming behind you to look at what you were doing / thinking about the functionality.

csilt commented:
> Is there any particular reason you don't actually have any commit message? Make it easier for the person coming behind you to look at what you were doing / thinking about the functionality.

I've updated the commit messages; let me know if you'd like any further changes.

donaldsharp commented on pimd/pim_util.c:
```diff
@@ -126,20 +128,90 @@ int pim_is_group_224_4(struct in_addr group_addr)
 }
 
-bool pim_is_group_filtered(struct pim_interface *pim_ifp, pim_addr *grp)
+static bool pim_cisco_match(const struct filter *filter, const struct in_addr *source,
```
Why isn't this commit broken up into multiple commits?  When I see `oh and I moved a bunch of stuff around` in the commit message and only see one commit it makes it much harder to review the changes for functionality -vs- the changes for structure.  Additionally it's hard to review and figure out what is going on.

ton31337 commented on pimd/pim_util.c:
```diff
@@ -126,20 +128,90 @@ int pim_is_group_224_4(struct in_addr group_addr)
-	return pl ? prefix_list_apply_ext(pl, NULL, &grp_pfx, true) ==
-			       PREFIX_DENY
-		  : false;
+	acl = access_list_lookup(PIM_AFI, pim_ifp->boundary_acl);
```
We always do this access_list_lookup() (even in a loop somewhere) which leads to unnecessary calls. Can we do this outside pim_is_group_filtered() (or store an access_list pointer instead of char)?

csiltala replied:
Converted to storing pointers for both prefix-list and access-list

ton31337 commented on tests/topotests/pim_basic_acl/test_pim_basic_acl.py:
```diff
@@ -0,0 +1,430 @@
+#
+# test_pim.py
+#
+# Copyright (c) 2018 Cumulus Networks, Inc.
```
Most likely this is not Cumulus?

csiltala replied:
Fixed this and test name

ton31337 commented on tests/topotests/pim_basic_acl/test_pim_basic_acl.py:
```diff
@@ -0,0 +1,430 @@
+#
+
+"""
+test_pim.py: Test pim
```
Ditto.

ton31337 commented on tests/topotests/pim_basic_acl/test_pim_basic_acl.py:
```diff
@@ -0,0 +1,430 @@
+    assert result is None, "Expected no IGMP sources other than for AutoRP Discovery"
+
+    # Send IGMP join from r2, check if r1 has IGMP source
+    tgen.routers()["r2"].cmd(("vtysh -c 'conf t' -c 'int {}' -c 'ip pim' -c 'ip igmp join {}'").format("r2-eth0", ASM_GROUP))
```
Please use vtysh_cmd()...

csiltala replied:
Done

ton31337 commented on tests/topotests/pim_basic_acl/test_pim_basic_acl.py:
```diff
@@ -0,0 +1,430 @@
+    test_func = partial(
+        topotest.router_json_cmp, r1, "show ip igmp sources json", expected
+    )
+    _, result = topotest.run_and_expect(test_func, None, count=5, wait=1)
```
Use much higher timers (minimum 15 seconds).

csiltala replied:
Set all timers that were lower than 20 to 20

ton31337 commented on tests/topotests/pim_basic_acl/test_pim_basic_acl.py:
```diff
@@ -0,0 +1,430 @@
+    test_func = partial(
+        topotest.router_json_cmp, r2, "show ip igmp join json", expected
+    )
+    _, result = topotest.run_and_expect(test_func, None, count=10, wait=1)
```
Ditto.

ton31337 commented on tests/topotests/pim_basic_acl/test_pim_basic_acl.py:
```diff
@@ -0,0 +1,430 @@
+    _, result = topotest.run_and_expect(test_func, None, count=10, wait=1)
+    assert result is None, "Expected IGMP join to be present but is absent"
+
+    tgen.routers()["r2"].cmd(("vtysh -c 'conf t' -c 'int {}' -c 'no ip igmp join {}' -c 'ip multicast boundary oil pim-oil-plist' -c 'ip igmp join {}'").format("r2-eth0", ASM_GROUP, ASM_GROUP))
```
Ditto.

ton31337 commented on tests/topotests/pim_basic_acl/test_pim_basic_acl.py:
```diff
@@ -0,0 +1,430 @@
+    test_func = partial(
+        topotest.router_json_cmp, r2, "show ip igmp join json", expected
+    )
+    _, result = topotest.run_and_expect(test_func, None, count=10, wait=1)
```
Ditto.

ton31337 commented on pimd/pim_join.c:
```diff
@@ -287,8 +287,8 @@ int pim_joinprune_recv(struct interface *ifp, struct pim_neighbor *neigh,
-			/* if we are filtering this group, skip the join */
-			if (filtered)
+			/* if we are filtering this group or (S,G), skip the join */
+			if (group_filtered || pim_is_group_filtered(pim_ifp, &sg.grp, &sg.src))
```
Here again, pim_is_group_filtered() is called in a loop where prefix-list/access-list lookup is done.

csiltala committed:
pimd,yang: Extend multicast boundary functionality

Add new interface command ip multicast boundary ACCESSLIST4_NAME. This
allows filtering on both source and group using the extended access-list
syntax vs. group-only as with the existing "ip multicast boundary oil"
command, which uses prefix-lists. If both are configured, the prefix-
list is evaluated first. The default behavior for both prefix-lists and
access-lists remains "deny", so the prefix-list must have a terminating
"permit" statement in order to also evaluate against the access-list.

The following example denies groups in range 229.1.1.0/24 and groups in
range 232.1.1.0/24 with source 10.0.20.2:

!
ip prefix-list pim-oil-plist seq 10 deny 229.1.1.0/24
ip prefix-list pim-oil-plist seq 20 permit any
!
access-list pim-acl seq 10 deny ip host 10.0.20.2 232.1.1.0 0.0.0.255
access-list pim-acl seq 20 permit ip any any
!
interface r1-eth0
 ip address 10.0.20.1/24
 ip igmp
 ip pim
 ip multicast boundary oil pim-oil-plist
 ip multicast boundary pim-acl
!

Signed-off-by: Corey Siltala <csiltala@atcorp.com>

csiltala committed:
pimd: Move ACL handling to pim_util.c

Move the extended access-list handling from pim_msdp_packet.c to
pim_util.c to allow use elsewhere in the daemon.

Signed-off-by: Corey Siltala <csiltala@atcorp.com>

csiltala committed:
pimd: Convert boundary_oil_plist to struct prefix_list

Rather than storing the prefix-list name and looking it up every time we use it, store a pointer to the prefix-list itself.

Signed-off-by: Corey Siltala <csiltala@atcorp.com>

csiltala commented on tests/topotests/pim_basic_acl/r3/frr.conf:
```diff
@@ -0,0 +1,13 @@
+hostname r3
+!
+! Need a pim statement in order to load pimd after merging pimd.conf
+debug pim events
```
Commented

csiltala commented on pimd/pim_iface.h:
```diff
@@ -133,8 +133,10 @@ struct pim_interface {
+	/* boundary prefix-list (group) */
 	char *boundary_oil_plist;
+	/* boundary access-list (source and group) */
+	char *boundary_acl;
```
I was following the existing `boundary_oil_plist` example but I've now converted both to `struct prefix_list *`/`struct access_list *`

csiltala commented on pimd/pim_igmpv3.c:
```diff
@@ -1823,7 +1826,6 @@ static bool igmp_pkt_grp_addr_ok(struct interface *ifp, const char *from_str,
 	 * If we receive a igmp report with the group in 224.0.0.0/24
 	 * then we should ignore it
 	 */
-
```
Fixed

csiltala commented on pimd/pim_iface.c:
```diff
@@ -1258,6 +1259,12 @@ static int gm_join_sock(const char *ifname, ifindex_t ifindex,
 	int join_fd;
 
+	if (pim_is_group_filtered(pim_ifp, &group_addr, &source_addr)) {
+		zlog_debug("%s: join failed for (S,G)=(%pPAs,%pPAs) due to multicast boundary filtering",
```
Fixed

csiltala commented on pimd/pim_cmd.c:
```diff
@@ -5871,6 +5871,21 @@ DEFUN(interface_no_ip_pim_boundary_oil,
+DEFPY_YANG(interface_ip_pim_boundary_acl,
+      interface_ip_pim_boundary_acl_cmd,
+      "[no] ip multicast boundary ACCESSLIST4_NAME$name",
+	  NO_STR
```
Fixed

ton31337 commented on tests/topotests/pim_boundary_acl/test_pim_boundary_acl.py:
```diff
@@ -0,0 +1,525 @@
+              no ip igmp join {} 10.0.20.2
+        """
+    ).format(SSM_GROUP))
+    tgen.routers()["r2"].cmd(("vtysh -c 'conf t' -c 'int {}' -c 'no ip igmp join {} 10.0.20.2'").format("r2-eth0", SSM_GROUP))
```
Any reason you don't use here as previously `r2.vtysh_cmd()`?

csiltala replied:
Had converted to `vtysh_cmd()` but missed removing this equivalent, fixed now.

ton31337 reviewed (approved):
LGTM

csiltala commented:
> LGTM

I just noticed an incorrectly copy-pasted comment and removed that (had same note about json_cmp wildcard for both ASM and SSM, inaccurate for SSM)
```
# Note: json_cmp treats "*" as wildcard but in this case that's actually what the source is
```
Only difference other than rebase on latest

Jafaral commented:
@greptile review this PR. take into account the current state of the code , so don't flag things that are already fixed on the master branch
